# bionetworks
Materials for semester project in Bioinformatics Institute (spring 2021)

## Building gene regulatory networks with structural properties
### Abstract
Gene regulatory network (GRN) is the model which is used for the description of gene expression levels complex interconnection on the whole-organism scale. It is represented as a graph that can be inferred from expression data, but the inference quality can not be properly assessed without known true structure which is the case in most real-world tasks. The answer is to challenge inference algorithms by artificial expression data generated from the random but still plausible simulated network.

In this work, the original network generation algorithm was implemented using Python. The algorithm was designed to capture the main local structural features of the real GRNs: 1) feed-forward loops (FFL) abundance comparing with randomized networks of the same structure; 2) vertex-based FFL network is the connected graph. Addressing these properties, at the first stage of generation, the whole FFL is attached until a predefined network size is reached. The classic preferential attachment algorithm is employed for network finishing afterwards. The resulting artificial GRN has a motif number distribution similar to the real GRNs' ones and captures the specific high FFL abundance. One is able to tune the size and sparsity of generated network using parameters.
### Objective
Random but plausible (in sense of local structure) artificial gene regulatory network (GRN) generation 
### Plan
1. Investigate real well-validated GRNs local structure
2. Implement generation algorithm

### Content

__./scripts/functions.py__ - the main code source __with comments__, diverse library of custom functions to work with GRNs

__./requirents.txt__ - prerequisites & dependencies: __pip install -r requirements.txt__

Project check-points in jupyter-notebooks:

__1_network_construction.ipynb__ - script converting raw data about network to binary matrix. Resulting matricies are stored in __networks__ folder in .gz format

__2.1_motif_search_algo.ipynb__ - launching motif search pipeline on diffent networks
![](pics/triads.png)

__2.2_motif_search_vis.ipynb__ - visualization of search results

__3.1_enrichment_prep.ipynb__ - test shuffling, QA and optimal swaps number estimation

__3.2_enrichment_analysis.ipynb__ - main results for enrichment analysis: plots, p-values and Z-scores
![](pics/enrichment.png)

__4_generation.ipynb__ - network generation algorithm
![](pics/algorithm.png)

# Raw data (bionetworks/raw_data)

__bionetworks/raw_data/GS_N800__ - network genetated by previous version of the tool at different noise levels

We used transcriptional network for yeast (__bionetworks/raw_data/tfcomb/tnet.txt__) from:
Balaji, S., Babu, M. M., Iyer, L. M., Luscombe, N. M., & Aravind, L. (2006). Comprehensive Analysis of Combinatorial
Regulation using the Transcriptional Regulatory Network of Yeast. Journal of Molecular Biology, 360(1), 213–227.
doi:10.1016/j.jmb.2006.04.029 (https://doi.org/10.1016/j.jmb.2006.04.029)

For _E.coli_ we used network (__bionetworks/raw_data/ecoli/network_tf_gene.txt) from:
Santos-Zavaleta, A., Salgado, H., Gama-Castro, S., Sánchez-Pérez, M., Gómez-Romero, L., Ledezma-Tejeida, D., …
Collado-Vides, J. (2018). RegulonDB v 10.5: tackling challenges to unify classic and high throughput knowledge of gene
regulation in E. coli K-12. Nucleic Acids Research. doi:10.1093/nar/gky1077 (https://doi.org/10.1093/nar/gky1077)

# Formatted networks & intermediate results (bionetworks/networks)

Results for all analysed networks: 
- Small random network for testing (__bionetworks/networks/test__)
- Yeast Tnet (__bionetworks/networks/yeast__)
- E.coli Tnet (__bionetworks/networks/ecoli__)
- Artificial networks from previous version of tool (__bionetworks/networks/gs1\0.1\0.01__)
- Resulting artificial network generated by novel algorithm (__bionetworks/network/artificial__)

All files are gzipped and named in accordance with the same pattern:
- __\*\_matrix.gz__ - interaction matrix defining the network, 1 at i-th row and j-th column is for there is link from gene j to gene i
- __counter.json__ - dictionary {motif type, string: number, integer}, result of motif counting for this network
- __motifs.gz__ - found motifs collection, {motif type, string: \[list of triads "{node\_x}\_{node\_y}\_{node\_z}"]}
- __shuffled_counters.gz__ - counters for shuffled iteraction matricies needed for motif enrichment analysis

